File: iteration2/CSE3063F21P1_GRP16/src/main/java/Models/Student.java
package Models;

import Services.CourseRegistrationSystem;
import Types.SemesterName;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Random;
import java.util.stream.Collectors;

public class Student extends Person {

    private StudentID studentID;
    private Advisor advisor;
    private StudentUtils studentUtils;
    private Transcript transcript;
    private Payment paymentAmount;
    private int StudentSemesterNo;
    private static FileWriter file;

    public Student(String newFirstName, String newLastName, ArrayList<Course> allCourses) {

        studentID = new StudentID("CSE", Integer.toString((int) Math.max(18, Math.random() * 20)), Integer.toString(Math.max(10, ((int) (Math.random() * 50)))));
        StudentSemesterNo = (int) (Math.random() * 5 + 2);
        FirstName = newFirstName;
        LastName = newLastName;
        transcript = new Transcript(this, allCourses);
        studentUtils=new StudentUtils();
    }

    public void saveToJson() {

        JSONObject obj = new JSONObject();
        obj.put("Name", FirstName);
        obj.put("LastName", LastName);

        JSONObject Trans = new JSONObject();
        Trans.put("gpa", transcript.getGPA());

        JSONArray passedCourses = new JSONArray(transcript.getPassedCourses().stream()
                .map(Course::getCourseCode)
                .collect(Collectors.toSet())
        );
        JSONArray failedCourses = new JSONArray(transcript.getFailedCourses().stream()
                .map(Course::getCourseCode)
                .collect(Collectors.toSet())
        );
        JSONArray selectedCourses = new JSONArray(getStudentUtils().getSelectedCourses().stream()
                .map(Course::getCourseCode)
                .collect(Collectors.toSet())
        );

        obj.put("PassedCourses", passedCourses);
        obj.put("FailedCourses", failedCourses);
        obj.put("SelectedCourses", selectedCourses);

        obj.put("TranskriptObj", Trans);
        try {

            file = new FileWriter("src/students/" + studentID.getStudentString() + ".json");
            file.write(obj.toString(4));
            // System.out.println("json created");

        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                file.flush();
                file.close();
            } catch (IOException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }
    }

    public StudentID getStudentID() {
        return studentID;
    }

    public void setStudentID(StudentID studentID) {
        this.studentID = studentID;
    }

    public Advisor getAdvisor() {
        return advisor;
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
    }

    public Transcript getTranscript() {
        return transcript;
    }

    public void setTranscript(Transcript transcript) {
        this.transcript = transcript;
    }

    public Payment getPaymentAmount() {
        return paymentAmount;
    }

    public void setPaymentAmount(Payment paymentAmount) {
        this.paymentAmount = paymentAmount;
    }

    public StudentUtils getStudentUtils() {
        return studentUtils;
    }

    public void setStudentUtils(StudentUtils studentUtils) {
        this.studentUtils = studentUtils;
    }

    public void addSemester(ArrayList<Course> Courses, Integer semesterNo, SemesterName semesterName){
        transcript.addSemester(this, Courses, semesterNo, semesterName);
    }
}

File: iteration3/CSE3063F21P1_GRP16/models/Student.py
import json
import random
from typing import List


from CSE3063F21P1_GRP16.models.Course import Course
from CSE3063F21P1_GRP16.models.StudentUtils import StudentUtils


class Student:
    def __init__(self, new_first_name: str, new_last_name: str, all_courses: List[Course]):
        self._student_id = StudentID("CSE", str(int(max(18, random.random() * 20))), str(int(max(10, int(random.random() * 50)))))
        self._student_semester_no = int(random.random() * 5 + 2)
        self._first_name = new_first_name
        self._last_name = new_last_name
        self._transcript = Transcript(self, all_courses)
        self._student_utils = StudentUtils()

    def save_to_json(self):
        obj = {
            "Name": self._first_name,
            "LastName": self._last_name
        }

        trans = {
            "gpa": self._transcript.get_gpa()
        }

        passed_courses = [c.get_course_code() for c in self._transcript.get_passed_courses()]
        failed_courses = [c.get_course_code() for c in self._transcript.get_failed_courses()]
        selected_courses = [c.get_course_code() for c in self.get_student_utils().get_selected_courses()]

        obj["PassedCourses"] = passed_courses
        obj["FailedCourses"] = failed_courses
        obj["SelectedCourses"] = selected_courses
        obj["TranskriptObj"] = trans

        with open(f"src/students/{self._student_id.get_student_string()}.json", "w") as file:
            file.write(json.dumps(obj, indent=4))

    def get_advisor(self):
        return self.advisor

    def set_advisor(self, advisor):
        self.advisor = advisor

    def get_transcript(self):
        return self._transcript

    def set_transcript(self, transcript):
        self._transcript = transcript

    def get_student_utils(self):
        return self._student_utils

    def set_student_utils(self, student_utils):
        self._student_utils = student_utils

    def add_semester(self, courses: List[Course], semester_no: int, semester_name: SemesterName):
        self._transcript.add_semester(self, courses, semester_no, semester_name)